File: test/Cartesian2d.py

import GMatTensor.Cartesian2d as GMat
import numpy as np

def ISCLOSE(a, b):
  assert np.abs(a-b) < 1.e-12

def A4_ddot_B2(A, B):
    return np.einsum('ijkl,lk->ij', A, B)

# Id - Tensor2

A = np.random.random([2, 2])
I = GMat.I2()
Id = GMat.I4d()
Is = GMat.I4s()
A = A4_ddot_B2(Is, A)
assert np.allclose(A4_ddot_B2(Id, A), A - GMat.Hydrostatic(A) * I)

# Deviatoric - Tensor2

A = np.random.random([2, 2])
B = np.array(A, copy=True)
tr = B[0, 0] + B[1, 1]
B[0, 0] -= 0.5 * tr
B[1, 1] -= 0.5 * tr
assert np.allclose(GMat.Deviatoric(A), B)

# Material points

K = 12.3
G = 45.6

gamma = 0.02
epsm = 0.12

Eps = np.array(
    [[epsm, gamma],
     [gamma, epsm]])

ISCLOSE(float(GMat.Epsd(Eps)), gamma)

# Elastic - stress

mat = GMat.Elastic(K, G)
mat.setStrain(Eps)
Sig = mat.Stress()

ISCLOSE(Sig[0,0], K * epsm)
ISCLOSE(Sig[1,1], K * epsm)
ISCLOSE(Sig[0,1], G * gamma)
ISCLOSE(Sig[1,0], G * gamma)

# Cusp - stress

mat = GMat.Cusp(K, G, [0.01, 0.03, 0.10])
mat.setStrain(Eps)
Sig = mat.Stress()

ISCLOSE(Sig[0,0], K * epsm)
ISCLOSE(Sig[1,1], K * epsm)
ISCLOSE(Sig[0,1], G * 0.0)
ISCLOSE(Sig[1,0], G * 0.0)
ISCLOSE(mat.epsp(), 0.02)
ISCLOSE(mat.currentIndex(), 1)

# Smooth - stress

mat = GMat.Smooth(K, G, [0.01, 0.03, 0.10])
mat.setStrain(Eps)
Sig = mat.Stress()

ISCLOSE(Sig[0,0], K * epsm)
ISCLOSE(Sig[1,1], K * epsm)
ISCLOSE(Sig[0,1], G * 0.0)
ISCLOSE(Sig[1,0], G * 0.0)
ISCLOSE(mat.epsp(), 0.02)
ISCLOSE(mat.currentIndex(), 1)

# Array2d

nelem = 3
nip = 2
mat = GMat.Array2d([nelem, nip])

I = np.zeros([nelem, nip], dtype='int')
I[0,:] = 1
mat.setElastic(I, K, G)

I = np.zeros([nelem, nip], dtype='int')
I[1,:] = 1
mat.setCusp(I, K, G, 0.01 + 0.02 * np.arange(100))

I = np.zeros([nelem, nip], dtype='int')
I[2,:] = 1
mat.setSmooth(I, K, G, 0.01 + 0.02 * np.arange(100))

eps = np.zeros((nelem, nip, 2, 2))

for e in range(nelem):
    for q in range(nip):
        fac = float((e + 1) * nip + (q + 1))
        eps[e, q, :, :] = fac * Eps

mat.setStrain(eps)
sig = mat.Stress()
epsp = mat.Epsp()

for q in range(nip):

    e = 0
    fac = float((e + 1) * nip + (q + 1))
    ISCLOSE(sig[e, q, 0, 0], fac * K * epsm)
    ISCLOSE(sig[e, q, 1, 1], fac * K * epsm)
    ISCLOSE(sig[e, q, 0, 1], fac * G * gamma)
    ISCLOSE(sig[e, q, 1, 0], fac * G * gamma)
    ISCLOSE(epsp[e, q], 0.0)

    e = 1
    fac = float((e + 1) * nip + (q + 1))
    ISCLOSE(sig[e, q, 0, 0], fac * K * epsm)
    ISCLOSE(sig[e, q, 1, 1], fac * K * epsm)
    ISCLOSE(sig[e, q, 0, 1], 0.0)
    ISCLOSE(sig[e, q, 1, 0], 0.0)
    ISCLOSE(epsp[e, q], fac * gamma)

    e = 2
    fac = float((e + 1) * nip + (q + 1))
    ISCLOSE(sig[e, q, 0, 0], fac * K * epsm)
    ISCLOSE(sig[e, q, 1, 1], fac * K * epsm)
    ISCLOSE(sig[e, q, 0, 1], 0.0)
    ISCLOSE(sig[e, q, 1, 0], 0.0)
    ISCLOSE(epsp[e, q], fac * gamma)

print('All checks passed')

File: test/Cartesian2d.cpp

#include <catch2/catch.hpp>
#include <xtensor/xrandom.hpp>
#include <GMatTensor/Cartesian2d.h>

#define ISCLOSE(a,b) REQUIRE_THAT((a), Catch::WithinAbs((b), 1e-12));

namespace GM = GMatTensor::Cartesian2d;

// namespace Prod2d {

//     template <class T, class S>
//     S A4_ddot_B2(const T& A, const S& B)
//     {
//         S C = xt::empty<double>({2, 2});
//         C.fill(0.0);

//         for (size_t i = 0; i < 2; i++) {
//             for (size_t j = 0; j < 2; j++) {
//                 for (size_t k = 0; k < 2; k++) {
//                     for (size_t l = 0; l < 2; l++) {
//                         C(i, j) += A(i, j, k, l) * B(l, k);
//                     }
//                 }
//             }
//         }

//         return C;
//     }

// } // namespace Prod2d

TEST_CASE("GMatTensor::Cartesian2d", "Cartesian2d.h")
{

    // SECTION("I4")
    // {
    //     GM::Tensor2 A = xt::random::randn<double>({2, 2});
    //     GM::Tensor4 I = GM::I4();
    //     REQUIRE(xt::allclose(Prod2d::A4_ddot_B2(I, A), A));
    // }

    // SECTION("I4s")
    // {
    //     GM::Tensor2 A = xt::random::randn<double>({2, 2});
    //     GM::Tensor4 Is = GM::I4s();
    //     REQUIRE(xt::allclose(Prod2d::A4_ddot_B2(Is, A), 0.5 * (A + xt::transpose(A))));
    // }

    // SECTION("I4d")
    // {
    //     GM::Tensor2 A = xt::random::randn<double>({2, 2});
    //     GM::Tensor2 I = GM::I2();
    //     GM::Tensor4 Id = GM::I4d();
    //     GM::Tensor2 B = 0.5 * (A + xt::transpose(A));
    //     REQUIRE(xt::allclose(Prod2d::A4_ddot_B2(Id, A), B - GM::Hydrostatic(B) * I));
    // }

    // SECTION("Deviatoric - Tensor2")
    // {
    //     GM::Tensor2 A = xt::random::randn<double>({2, 2});
    //     GM::Tensor2 B = A;
    //     double tr = B(0, 0) + B(1, 1);
    //     B(0, 0) -= 0.5 * tr;
    //     B(1, 1) -= 0.5 * tr;
    //     REQUIRE(xt::allclose(GM::Deviatoric(A), B));
    // }

    SECTION("Deviatoric - List")
    {
        GM::Tensor2 A = xt::random::randn<double>({2, 2});
        GM::Tensor2 B = A;
        double tr = B(0, 0) + B(1, 1);
        B(0, 0) -= 0.5 * tr;
        B(1, 1) -= 0.5 * tr;
        auto M = xt::xtensor<double,3>::from_shape({3, 2, 2});
        auto R = xt::xtensor<double,3>::from_shape(M.shape());
        for (size_t i = 0; i < M.shape(0); ++i) {
            xt::view(M, i, xt::all(), xt::all()) = static_cast<double>(i) * A;
            xt::view(R, i, xt::all(), xt::all()) = static_cast<double>(i) * B;
        }
        REQUIRE(xt::allclose(GM::Deviatoric(M), R));
    }

    SECTION("Deviatoric - Matrix")
    {
        GM::Tensor2 A = xt::random::randn<double>({2, 2});
        GM::Tensor2 B = A;
        double tr = B(0, 0) + B(1, 1);
        B(0, 0) -= 0.5 * tr;
        B(1, 1) -= 0.5 * tr;
        auto M = xt::xtensor<double,4>::from_shape({3, 4, 2, 2});
        auto R = xt::xtensor<double,4>::from_shape(M.shape());
        for (size_t i = 0; i < M.shape(0); ++i) {
            for (size_t j = 0; j < M.shape(1); ++j) {
                xt::view(M, i, j, xt::all(), xt::all()) = static_cast<double>(i * M.shape(1) + j) * A;
                xt::view(R, i, j, xt::all(), xt::all()) = static_cast<double>(i * M.shape(1) + j) * B;
            }
        }
        REQUIRE(xt::allclose(GM::Deviatoric(M), R));
    }

    // SECTION("Hydrostatic - Tensor2")
    // {
    //     GM::Tensor2 A = xt::random::randn<double>({2, 2});
    //     A(0, 0) = 1.0;
    //     A(1, 1) = 1.0;
    //     REQUIRE(GM::Hydrostatic(A)() == Approx(1.0));
    // }

    SECTION("Hydrostatic - List")
    {
        GM::Tensor2 A = xt::random::randn<double>({2, 2});
        A(0, 0) = 1.0;
        A(1, 1) = 1.0;
        auto M = xt::xtensor<double,3>::from_shape({3, 2, 2});
        auto R = xt::xtensor<double,1>::from_shape({M.shape(0)});
        for (size_t i = 0; i < M.shape(0); ++i) {
            xt::view(M, i, xt::all(), xt::all()) = static_cast<double>(i) * A;
            R(i) = static_cast<double>(i);
        }
        REQUIRE(xt::allclose(GM::Hydrostatic(M), R));
    }

    SECTION("Hydrostatic - Matrix")
    {
        GM::Tensor2 A = xt::random::randn<double>({2, 2});
        A(0, 0) = 1.0;
        A(1, 1) = 1.0;
        auto M = xt::xtensor<double,4>::from_shape({3, 4, 2, 2});
        auto R = xt::xtensor<double,2>::from_shape({M.shape(0), M.shape(1)});
        for (size_t i = 0; i < M.shape(0); ++i) {
            for (size_t j = 0; j < M.shape(1); ++j) {
                xt::view(M, i, j, xt::all(), xt::all()) = static_cast<double>(i * M.shape(1) + j) * A;
                R(i, j) = static_cast<double>(i * M.shape(1) + j);
            }
        }
        REQUIRE(xt::allclose(GM::Hydrostatic(M), R));
    }

    // SECTION("Equivalent_deviatoric - Tensor2")
    // {
    //     GM::Tensor2 A = xt::zeros<double>({2, 2});
    //     A(0, 1) = 1.0;
    //     A(1, 0) = 1.0;
    //     REQUIRE(GM::Equivalent_deviatoric(A)() == Approx(std::sqrt(2.0)));
    // }

    SECTION("Equivalent_deviatoric - List")
    {
        GM::Tensor2 A = xt::zeros<double>({2, 2});
        A(0, 1) = 1.0;
        A(1, 0) = 1.0;
        auto M = xt::xtensor<double,3>::from_shape({3, 2, 2});
        auto R = xt::xtensor<double,1>::from_shape({M.shape(0)});
        for (size_t i = 0; i < M.shape(0); ++i) {
            xt::view(M, i, xt::all(), xt::all()) = static_cast<double>(i) * A;
            R(i) = static_cast<double>(i) * std::sqrt(2.0);
        }
        REQUIRE(xt::allclose(GM::Equivalent_deviatoric(M), R));
    }

    SECTION("Equivalent_deviatoric - Matrix")
    {
        GM::Tensor2 A = xt::zeros<double>({2, 2});
        A(0, 1) = 1.0;
        A(1, 0) = 1.0;
        auto M = xt::xtensor<double,4>::from_shape({3, 4, 2, 2});
        auto R = xt::xtensor<double,2>::from_shape({M.shape(0), M.shape(1)});
        for (size_t i = 0; i < M.shape(0); ++i) {
            for (size_t j = 0; j < M.shape(1); ++j) {
                xt::view(M, i, j, xt::all(), xt::all()) = static_cast<double>(i * M.shape(1) + j) * A;
                R(i, j) = static_cast<double>(i * M.shape(1) + j) * std::sqrt(2.0);
            }
        }
        REQUIRE(xt::allclose(GM::Equivalent_deviatoric(M), R));
    }

}
